File: Algorithms/Python3/00020-Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        closedMap = {
            ")" : "(",
            "}" : "{",
            "]" : "["
        }

        for c in s:
            if c in closedMap:
                if stack and stack[-1] == closedMap[c]:
                    stack.pop()
                else: 
                    return False
            else:
                stack.append(c)
        
        return not stack

'''
stack

"(({{[[)]}}))"
1) stack is empty -> string is valid
2) right bracket doesnt correspond to left bracket -> string is invalid

if stack is empty at end, return true

use a dictionary / map to map closed brackets to open brackets
'''

File: Algorithms/C++/00020-Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        std::stack<char> stack;
        for(const auto& c : s){
            if(c == '(' || c == '{' || c == '['){
                stack.push(c);
                continue;
            }

            if(stack.empty()) return false;
            int top = stack.top();

            if(top == '(' && c == ')') stack.pop();
            else if(c == ')') stack.push(c);

            if(top == '{' && c == '}') stack.pop();
            else if(c == '}') stack.push(c);

            if(top == '[' && c == ']') stack.pop();
            else if(c == ']') stack.push(c);
           
        }

        return stack.empty();
    }
};

/*

    
    algorithm
    
    if we see a open bracket, we will push onto stack
    if we see a closed bracket (of same type) we will pop open bracket

    if string is valid, then stack should be empty

*/